### PR TITLE
Just to see the diff

### PR DESCRIPTION
**Pull request by Muennighoff**

### CODE REVIEW

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 1]
```diff
@@ -263,11 +263,38 @@ def get_lr():
...
+
 for step, batch in enumerate(train_dataloader, start=1):
     if args.resume_from_checkpoint and step < resume_step:
         continue  # we need to skip steps until we reach the resumed step
+    if args.selection_method:
+        if args.selection_method == "uniform":
+            batch = {k: v[:args.train_batch_size_select] for k,v in batch.items()}
+        elif args.selection_method == "rholoss":
+            with torch.no_grad():
+                out = model(batch, labels=batch, use_cache=False).loss
```
Comment (loubnabnl) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:282]: Comment about the batch size:  we're assuming that we can fit a batch size of 320 with our workers, but I think we can only fit 12 sequences on A100 40GB (so on 16 workers: batch of 16*12=192). 

So we should probably either incorporate gradient accumulation and store the losses for 2 iterations (2 * 10 (small bz) * 16 gpus=320) or we can change the batch sizes from 320/32 to something that suits us with a 10% ratio like 160/16. In the paper they just talk about the 10% ratio but I'm not sure if using large batches si also important?

**Review by loubnabnl — COMMENTED** [review]
Looks great! I just left a comment on the order of the losses and the batch size we can fit

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py (hunk 1 above)
Comment (Muennighoff) [reply examples/research_projects/codeparrot/scripts/codeparrot_training.py:288]: There are no gradients here which means that a) we can likely fit a bigger batch size than 12 b) instead of grad acc. we can just run multiple times right after another & store the losses if it doesnt fit

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 2]
```diff
@@ -263,11 +263,38 @@ def get_lr():
 completed_steps = 0
 t_start = time.time()
 loss_tracking = 0
+
+if args.compute_irred_losses:
+    irred_losses = torch.zeros(len(train_dataloader) * args.train_batch_size, device="cpu")
+elif args.irred_losses and not args.compute_irred_losses:
+    # Should be of shape len(train_dataloader) * args.train_batch_size
+    irred_losses = torch.load(args.irred_losses, map_location=torch.device("cpu"))
```
[the remark this answers is not in the document]
Comment (Muennighoff) [reply examples/research_projects/codeparrot/scripts/codeparrot_training.py:277]: > saving the losses with the sequences

You mean saving a mapping from input_ids to loss?

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py (hunk 1 above)
Comment (loubnabnl) [reply examples/research_projects/codeparrot/scripts/codeparrot_training.py:288]: yes right! by grad acc. I also meant doing similar iterations over the losses

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (loubnabnl) [reply examples/research_projects/codeparrot/scripts/codeparrot_training.py:277]: yes like adding an index to each sequence and saving it with the losses

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 3]
```diff
@@ -119,20 +119,29 @@ def setup_logging(args):
         transformers.utils.logging.set_verbosity_error()
     return logger, run_name
 
-
 def create_dataloaders(args):
-    ds_kwargs = {"streaming": True}
+    ds_kwargs = {"streaming": args.no_streaming}
     train_data = load_dataset(args.dataset_name_train, split="train", **ds_kwargs)
```
Comment (loubnabnl) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:124]:
```suggestion
    train_data = load_dataset(args.dataset_name_train, split="train", use_auth_token=True)
```

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 4]
```diff
@@ -119,20 +119,29 @@ def setup_logging(args):
         transformers.utils.logging.set_verbosity_error()
     return logger, run_name
 
-
 def create_dataloaders(args):
-    ds_kwargs = {"streaming": True}
+    ds_kwargs = {"streaming": args.no_streaming}
```
Comment (loubnabnl) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:123]: I think we can remove streaming option all together since we're not using it and there'll be no need to separate split method
```suggestion
```

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 5]
```diff
@@ -119,20 +119,29 @@ def setup_logging(args):
...
+    if args.dataset_name_train == args.dataset_name_valid:
+        # Split the dataset into train and validation
+        if args.no_streaming:
+            train_data = train_data.shuffle()
+            valid_data = train_data.select(range(int(len(train_data) * 0.05)))
+            train_data = train_data.select(range(int(len(train_data) * 0.05), len(train_data)))
+        else:
+            valid_data = train_data.take(5000)
+    else:
+        valid_data = load_dataset(args.dataset_name_valid, split="train", **ds_kwargs)
```
[the remark this answers is not in the document]
Comment (Muennighoff) [reply examples/research_projects/codeparrot/scripts/codeparrot_training.py:135]: I think   `.train_test_split(` does not work for streaming datasets, but we can use it if not streaming

**Comment by loubnabnl** [comment]
Summary of the changes I added:
* fix device & size mismathes, (e.g len(dataloader) doesn't work on iterable dataset + take into account the number of processes)
* in `transformers` the forward pass of GPT2 returns the average loss over the entire batch, and the `loss.repeat(batch_size)` before calling `accelerate.gather`
was just repeating that average so would ended up selecting the same value. I changed this [line](https://github.com/huggingface/transformers/blob/799cea64ac1029d66e9e58f18bc6f47892270723/src/transformers/models/gpt2/modeling_gpt2.py#L1076) in 
 gpt2_modeling of `transformers` to add [reduction="none"](https://github.com/huggingface/transformers/blob/799cea64ac1029d66e9e58f18bc6f47892270723/src/transformers/models/gpt2/modeling_gpt2.py#L1076) (see requirements). This will return the loss of each token so average to get loss per sequence.
* added sanity checks on the order of irreducible losses (this requires using the same batch size.. during their computation and when loading them)
* after selecting a small batch size (32) the same batch would be in all workers so no need to gather values, one thing we will need to add is split the batch (32) to 2 or 3 equal chunks then do grad acc because 32 won't fit in one worker.

**Comment by Muennighoff** [comment]
> * after selecting a small batch size (32) the same batch would be in all workers so no need to gather values, one thing we will need to add is split the batch (32) to 2 or 3 equal chunks then do grad acc because 32 won't fit in one worker.

Amazing work - do you want me to add the last point you mentioned?

**Comment by loubnabnl** [comment]
You can add it you have time, otherwise I will add it later 🤗

**Comment by Muennighoff** [comment]
> You can add it you have time, otherwise I will add it later 🤗

Done, but not tested. May have a bug 👻
